base/ltproperties.dtx

```diff
@@ -0,0 +1,930 @@
+      {
+        \gdef \@multiplelabels
+          { \@latex@warning@no@line { There~were~multiply-defined~labels } }
+        \@latex@warning@no@line { Label~`#1'~multiply~defined }
+      }
+      {
+        \tl_new:c { r@#1 }
+        \tl_gset:cn { r@#1 }{#2}
+      }
+    \@@_data:nnn {#1} #2 { \q_recursion_tail } { ? } \q_recursion_stop
+  }
+\cs_new_protected:Npn \@@_data:nnn #1#2#3
+  {
+    \quark_if_recursion_tail_stop:n {#2}
+    \tl_gclear_new:c { g_@@_label_ \tl_to_str:n {#1} _ \tl_to_str:n {#2} _tl }
+    \tl_gset:cn { g_@@_label_ \tl_to_str:n {#1} _ \tl_to_str:n {#2} _tl } {#3}
+    \@@_data:nnn {#1}
+  }  
+%    \end{macrocode}
+%
+% This command is used in \cs{enddocument} to test if some label values haved changed.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@kernel@new@label@record@testdef #1 #2
+  {
+    \tl_if_eq:cnF { r@#1 } {#2} 
+      { \@tempswatrue }
+  }
+%    \end{macrocode}
+% \end{macro}
+% \end{macro}
+%
+% \subsection{Tests and warnings}
+%
+% \begin{macro}[pTF]{\property_if_exist:n}
+%   Tests if property has been declared.
+%    \begin{macrocode}
+\prg_new_conditional:Npnn \property_if_exist:n #1 { p , T , F,  TF }  
+  % #1 property
+  {
+    \cs_if_exist:cTF { @@_name_ #1 : }
+      {
+        \prg_return_true:
+      }
+      {
+        \prg_return_false:
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{IfPropertyExistTF}
+%    \begin{macrocode}
+\cs_new_eq:NN \IfPropertyExistTF \property_if_exist:nTF  
+%    \end{macrocode}
+% \end{macro}
+% 
+% \begin{macro}[pTF]{\property_if_recorded:n}
+%   Tests if the label has been set.
+%   This can then be used to setup e.g.~rerun messages.
+%    \begin{macrocode}
+\prg_new_conditional:Npnn \property_if_recorded:n #1 { p , T , F,  TF }  
+  % #1 label 
+  {
+    \tl_if_exist:cTF { r@#1 }
+      {
+        \prg_return_true:
+      }
+      {
+        \prg_return_false:
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\IfLabelExistTF}
+%    \begin{macrocode}
+\cs_new_eq:NN \IfLabelExistTF \property_if_recorded:nTF  
+%    \end{macrocode}
+% \end{macro} 
+%
+% \begin{macro}[pTF]{\property_if_recorded:nn}
+% tests if the label/property combination has been set
+% This can then be used to setup e.g. rerun messages.
+%    \begin{macrocode}
+\prg_new_conditional:Npnn \property_if_recorded:nn #1#2 { p , T , F,  TF }  
+  % #1 label #2 property
+  {
+    \tl_if_exist:cTF { g_@@_label_ \tl_to_str:n {#1} _ \tl_to_str:n {#2} _tl }
+      {
+        \prg_return_true:
+      }
+      {
+        \prg_return_false:
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\IfPropertyRecordedTF}
+%    \begin{macrocode}
+\cs_new_eq:NN \IfPropertyRecordedTF \property_if_recorded:nnTF  
+%    \end{macrocode}
+% \end{macro}
+% 
+% \begin{macro}{\property_ref_undefined_warn:}
+%    \begin{macrocode}
+\cs_new_protected:Npn \property_ref_undefined_warn:
+ {
+   \flag_if_raised:nT { l_@@_ref_flag }
+    {
+      \G@refundefinedtrue
```

Comment:
Should this be a `\legacy_if_set_true:n`?

Reply:
(Also applies to the other uses)

Reply:
no `\G@refundefinedtrue` is command that defines a warning (and yes the name is confusing).

Reply:
Add a note to the source to highlight this?